tracy: fix build on macos (fixes #223469)

maxammann opened the pull request:
###### Description of changes
The UniformTypeIdentifiers is required since some tracy version. This patch fixes the build on macOS

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->


fixes https://github.com/NixOS/nixpkgs/issues/223469

maxammann committed:
tracy: fix build on macos (fixes #223469)

The UniformTypeIdentifiers is required for apple SDK 11+.
This patch fixes the build on newwer macOS versions.

maxammann commented:
@nagisa this fixes the build. I'm not sure though why I need to access that framework through the darwin symbol, whereas AppKit does not require that.

wegank commented:
`UniformTypeIdentifiers` is not available on macOS 10.12. Is there a way to build the package without this framework on `x86_64-darwin`?

maxammann commented:
> `UniformTypeIdentifiers` is not available on macOS 10.12. Is there a way to build the package without this framework on `x86_64-darwin`?

So on macOS 10.12 the framework is not required: https://github.com/wolfpld/tracy/blob/33a640f848c77c6297b0f04dc6567d59eaed321b/nfd/nfd_cocoa.m#L12-L29

Can we somehow only add this framework for SDKs < 11.0? Should I just assert on the architecture of darwin?

wegank commented:
You can do something like `lib.versionAtLeast stdenv.hostPlatform.darwinMinVersion "11"`.

maxammann commented:
Oke, so this should work now, both on x86 and on latest Apple Silicon.

wegank commented:
```
ld: framework not found UniformTypeIdentifiers
clang-11: error: linker command failed with exit code 1 (use -v to see invocation)
```

You might need to remove `-framework UniformTypeIdentifiers` somewhere in the code, conditionally.

maxammann committed:
tracy: update from 0.9 to 0.9.1

maxammann committed:
tracy: Remove UniformTypeIdentifiers depending on SDK

The UniformTypeIdentifiers is only required on SDK 11+. On SDK 10 (on which it is not available) we can remove it using a patch.

maxammann commented:
I added a patch :) let's try again!

wegank commented:
@ofborg build tracy tracy.passthru.tests

wegank commented:
> Should the linux tests succeed? I did not really touch Linux here

They should: https://hydra.nixos.org/job/nixpkgs/trunk/tracy.x86_64-linux

wegank commented:
Looks like `LEGACY=1` is applied in AUR: https://aur.archlinux.org/cgit/aur.git/commit/?h=tracy&id=a6bd906c2c7aee86bd5d0022c0c7cf2fdf7e4228

maxammann committed:
tracy: Fix compilation for 0.9.1

- Compile the profiler with LEGACY=1 to link against GLFW and use X11
- Remove the gtk3 dependency and add freetype2, dbus, and hicolor-icon-theme

maxammann commented:
Alright I blindly changed what was done on the AUR.

I dont have a linux machine with desktop available right now. So if somebody could test this, that would be great!

Else I can test in 2 months :)

wegank commented on pkgs/development/tools/tracy/default.nix:
```diff
@@ -1,25 +1,29 @@
     ++ lib.optionals stdenv.isDarwin [ Carbon AppKit freetype ]
-    ++ lib.optionals stdenv.isLinux [ gtk3 tbb dbus ];
+    ++ lib.optionals (stdenv.isDarwin && lib.versionAtLeast stdenv.hostPlatform.darwinMinVersion "11") [ darwin.apple_sdk.frameworks.UniformTypeIdentifiers ]
+    ++ lib.optionals stdenv.isLinux [ tbb dbus pkgs.hicolor-icon-theme freetype ];
```
`pkgs.hicolor-icon-theme` doesn't look right. Anyway, I can fix this for you tomorrow or next Monday.

maxammann replied:
Ah yeah, pkgs is not there :)

Thanks! Let's see if it works and if not, that would be great!

maxammann replied:
I upgraded directly to 0.9.1 so the patch is compatible.

wegank commented:
Result of `nixpkgs-review pr 223835` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>tracy</li>
  </ul>
</details>
